공지사항 댓글 등록 API 개발 및 공지사항 등록 리팩토링

zlzzlzz2l opened the pull request:
# 공지사항 댓글 등록 API 개발 및 공지사항 등록 리팩토링

---

- 공지사항 댓글 등록 API를 추가하였습니다
- 공지사항 등록 API를 수정하였습니다.

---

## 공지사항 댓글 등록

### 공지사항 댓글이 작성되었을 때
![image](https://user-images.githubusercontent.com/48669011/161106429-12556f48-437f-4233-af9b-42388f40cdef.png)

### 댓글에 " " or "" or null이 들어갔을 때
![image](https://user-images.githubusercontent.com/48669011/161106608-65ee944b-bb7f-4cbd-a092-810172bdce8e.png)

![image](https://user-images.githubusercontent.com/48669011/161106649-4139ce88-73d5-450c-99ae-f53e72934b06.png)

![image](https://user-images.githubusercontent.com/48669011/161106713-0e225a3b-3e06-4bea-a7b6-502c03d99918.png)

### 공지사항 데이터가 없을 때
![image](https://user-images.githubusercontent.com/48669011/161106978-bc848203-3c31-4f7b-bfc1-0239b533e95b.png)

zlzzlzz2l committed:
Update: #35 - 에러 메시지 추가

- 공지사항 데이터가 없을 때와 댓글이 작성되지 않았을 때의 에러 메시지를 추가하였습니다.

zlzzlzz2l committed:
Update: #35 - 에러 메시지 static 변수 추가

- 공지사항 데이터가 존재하지 않을 때와 댓글이 작성되지 않았을 때의 에러 메시지를 위해 추가하였습니다.

zlzzlzz2l committed:
Create: #35 - 공지사항 댓글 요청 DTO 생성

- 공지사항 댓글 등록 DTO 생성

zlzzlzz2l committed:
Create: #35 - 공지사항 댓글 컨트롤러 생성

- 공지사항 댓글 등록 메서드 생성

zlzzlzz2l committed:
Create: #35 - 서비스 레이어 공지사항 댓글 DTO 생성

- 공지사항 댓글 등록 DTO 생성

zlzzlzz2l committed:
Create: #35 - 공지사항 댓글 서비스 인터페이스 생성

- 공지사항 댓글 등록 메서드 선언부 추가
- 삭제 메서드 선언부 추가

zlzzlzz2l committed:
Create: #35 - 공지사항 댓글 서비스 구현 클래스 생성

- 공지사항 댓글 등록 메서드 작성
- 공지사항이 존재하지 않을 때의 예외처리 작성
- 삭제 메서드 작성

zlzzlzz2l committed:
Create: #35 - 공지사항 댓글 레포지토리 인터페이스 생성

- 식별자와 삭제 여부로 공지사항 댓글을 찾기 위해 추가하였습니다.

zlzzlzz2l committed:
Test: #35 - 공지사항 댓글 서비스 테스트 생성

- 공지사항 댓글 작성 성공 테스트 작성
- 공지사항 댓글 작성 예외 처리 테스트 작성

zlzzlzz2l committed:
Test: #35 - 공지사항 댓글 레포지토리 테스트 생성

- 공지사항 댓글 엔티티 등록 테스트 작성

zlzzlzz2l committed:
Refactor: #35 - 공지사항 요청 DTO 생성

- 공지사항 등록 DTO 생성

zlzzlzz2l committed:
Refactor: #35 - 공지사항 컨트롤러 수정

- 공지사항 등록 메서드 파라미터 수정

zlzzlzz2l committed:
Create: #35 - 서비스 레이어 공지사항 DTO 생성

zlzzlzz2l committed:
Refactor: #35 - 공지사항 서비스 인터페이스 수정

- 공지사항 등록 메서드 인자 수정

zlzzlzz2l committed:
Refactor: #35 - 공지사항 서비스 구현 클래스 수정

- 공지사항 등록 메서드 파라미터를 수정하면서 변수 이름도 변경하였습니다.

zlzzlzz2l committed:
Refactor: #35 - 공지사항 레포지토리 인터페이스 수정

- null 처리를 위해 Optional 클래스로 변경하였습니다.

zlzzlzz2l committed:
Refactor: #35 - 공지사항 서비스 테스트 수정

- 공지사항 등록 테스트 수정
- 관리자가 존재하지 않을 때의 예외 처리 추가

zlzzlzz2l committed:
Refactor: #35 - 공지사항 레포지토리 테스트 수정

0xe82de commented on src/main/java/com/fourfourfour/damoa/domain/notice/controller/dto/NoticeCommentRequestDto.java:
```diff
@@ -0,0 +1,38 @@
+    @NoArgsConstructor
+    public static class RegisterDto {
+
+        private Long noticeSeq;
```
데이터가 양수만 들어올 수 있도록 제한하면 좋을 것 같아요.

zlzzlzz2l replied:
반영했습니다!

0xe82de commented on src/main/java/com/fourfourfour/damoa/domain/notice/controller/NoticeCommentController.java:
```diff
@@ -0,0 +1,39 @@
+@Slf4j
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/notice/comment")
```
요청 URI를 `/api/v1/notices/{noticeSeq}/comments` 로 바꾸는 것은 어떤가요?

zlzzlzz2l replied:
반영했습니다!

0xe82de commented on src/test/java/com/fourfourfour/damoa/domain/notice/service/NoticeCommentServiceImplTest.java:
```diff
@@ -0,0 +1,135 @@
+    @Autowired
+    private PasswordEncoder passwordEncoder;
+
+    private MemberRequestDto.RegisterDto adminRegisterDto, memberRegisterDto;
+
+    private NoticeRequestDto.RegisterDto noticeRegisterDto;
```
앞으로 추가될 테스트 케이스에서 2개 이상의 테스트용 DTO가 필요하다면 변수명 끝에 숫자를 붙여도 좋을 것 같아요.
```text
adminRegisterDto => adminRegisterDto1
memberRegisterDto => memberRegisterDto1
noticeRegisterDto => noticeRegisterDto1
```

zlzzlzz2l replied:
반영했습니다!

0xe82de commented on src/test/java/com/fourfourfour/damoa/domain/notice/repository/NoticeCommentRepositoryTest.java:
```diff
@@ -0,0 +1,116 @@
+    @Autowired
+    private PasswordEncoder passwordEncoder;
+
+    private Member admin, member;
+
+    private Notice notice;
```
앞으로 추가될 테스트 케이스에서 2개 이상의 테스트용 DTO가 필요하다면 변수명 끝에 숫자를 붙여도 좋을 것 같아요.
```text
admin => admin1
member => member1
notice => notice1
```

zlzzlzz2l replied:
반영했습니다!

0xe82de commented on src/test/java/com/fourfourfour/damoa/domain/notice/service/NoticeServiceImplTest.java:
```diff
@@ -38,15 +40,20 @@
     private PasswordEncoder passwordEncoder;
 
-    private MemberRequestDto.RegisterDto registerDto1;
+    private MemberRequestDto.RegisterDto adminRegisterDto;
+
+    private NoticeRequestDto.RegisterDto noticeRegisterDto;
```
앞으로 추가될 테스트 케이스에서 2개 이상의 테스트용 DTO가 필요하다면 변수명 끝에 숫자를 붙여도 좋을 것 같아요.
```text
adminRegisterDto => adminRegisterDto1
noticeRegisterDto => noticeRegisterDto1
```

zlzzlzz2l replied:
반영했습니다!

0xe82de commented on src/test/java/com/fourfourfour/damoa/domain/notice/repository/NoticeRepositoryTest.java:
```diff
@@ -57,8 +58,8 @@ public void setUp() {
     @Test
-    @DisplayName("공지사항 엔티티 등록 테스트")
-    public void register() {
+    @DisplayName("공지사항 엔티티 등록")
+    public void noticeRegister() {
```
메서드가 정상적으로 처리되었을 때 테스트 케이스인 것 같아요. 아래처럼 수정하는 것은 어떨까요?
```java
@DisplayName("공지사항 엔티티 등록 : 성공")
public void noticeRegisterSuccess()
```

zlzzlzz2l replied:
넵 반영했습니다!

0xe82de reviewed (changes requested):
고생하셨습니다. :) 리뷰 확인해주시고 피드백 주세요.
감사합니다.

zlzzlzz2l committed:
Refactor: #35 - 공지사항 댓글 요청 DTO 수정

- 공지사항 식별자는 PathVariable을 사용하기 때문에 삭제하였습니다.
- PathVariable인 공지사항 식별자를 받기 위해 서비스 변환 DTO 메서드에 공지사항 식별자를 인자로 추가하였습니다.

zlzzlzz2l committed:
Refactor: #35 - 공지사항 댓글 컨트롤러 수정

- Valid 어노테이션 삭제 및 Validated 어노테이션 추가하였습니다.
- 요청 URI를 수정하였습니다.
- 공지사항 식별자를 공지사항 댓글 등록 메서드의 인자로 추가하였습니다.
- 로그 출력을 수정하였습니다.
- 서비스 변환 DTO 메서드에 공지사항 식별자를 추가하였습니다.

zlzzlzz2l committed:
Refactor: #35 - 공지사항 댓글 서비스 테스트 수정

- 2개 이상의 테스트용 DTO가 필요할 수도 있기 때문에 변수명 끝에 숫자를 추가하였습니다.
- 공지사항 댓글 요청 DTO에 공지사항 식별자를 선언하지 않았기 때문에 삭제하였습니다.
- 공지사항 댓글 요청 DTO를 서비스 변환 DTO로 변환할 때 공지사항 식별자가 필요하기 때문에 인자로 추가하였습니다.

zlzzlzz2l committed:
Refactor: #35 - 공지사항 댓글 레포지토리 수정

- 2개 이상의 테스트용 엔티티가 필요할 수도 있기 때문에 변수명 끝에 숫자를 추가하였습니다.

zlzzlzz2l committed:
Refactor: #35 - 공지사항 서비스 테스트 수정

- 2개 이상의 테스트용 DTO가 필요할 수도 있기 때문에 변수명 끝에 숫자를 추가하였습니다.

zlzzlzz2l committed:
Refactor: #35 - 공지사항 레포지토리 테스트 수정

- 테스트 케이스의 이름을 명확하게 나타내기 위해 수정하였습니다.

yjs1213 reviewed (approved):
고생하셨습니다

zlzzlzz2l committed:
Refactor: #35 - 공지사항 등록 컨트롤러 수정

- 요청 URI 수정
- 로그 출력 수정

zlzzlzz2l committed:
Refactor: #35 - 공지사항 컨트롤러 수정

- 요청 URI 수정

0xe82de reviewed (approved):
고생하셨습니다. :)
